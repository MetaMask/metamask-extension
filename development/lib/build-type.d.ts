import { Infer, Struct } from 'superstruct';

// TODO: Replace `any` with type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Unique<Element extends Struct<any>> = (
  struct: Struct<Infer<Element>[], Infer<Element>>,
  eq?: (a: Infer<Element>, b: Infer<Element>) => boolean,
) => Struct<Infer<Element>[], Infer<Element>>;
