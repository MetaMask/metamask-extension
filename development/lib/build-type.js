const fs = require('fs');
const { AssertionError } = require('assert');
const path = require('path');
const {
  object,
  string,
  record,
  optional,
  array,
  refine,
  any,
  boolean,
  coerce,
  union,
  number,
  unknown,
  validate,
  nullable,
  never,
  literal,
} = require('superstruct');
const yaml = require('yaml');
const { uniqWith } = require('lodash');

const BUILDS_YML_PATH = path.resolve('./builds.yml');

/**
 * @type {import('superstruct').Infer<typeof BuildTypesStruct> | null}
 */
let cachedBuildTypes = null;

/**
 * Ensures that the array item contains only elements that are distinct from each other
 *
 * @template {Struct<any>} Element
 * @type {import('./build-type').Unique<Element>}
 */
const unique = (struct, eq) =>
  refine(struct, 'unique', (value) => {
    if (uniqWith(value, eq).length === value.length) {
      return true;
    }
    return 'Array contains duplicated values';
  });

const EnvDefinitionStruct = coerce(
  object({ key: string(), value: unknown() }),
  refine(record(string(), any()), 'Env variable declaration', (value) => {
    if (Object.keys(value).length !== 1) {
      return 'Declaration should have only one property, the name';
    }
    return true;
  }),
  (value) => ({ key: Object.keys(value)[0], value: Object.values(value)[0] }),
);

const EnvArrayStruct = unique(
  array(union([string(), EnvDefinitionStruct])),
  (a, b) => {
    const keyA = typeof a === 'string' ? a : a.key;
    const keyB = typeof b === 'string' ? b : b.key;
    return keyA === keyB;
  },
);

/**
 * Ensures a number is within a given range
 *
 * @param {number} min
 * @param {number} max
 */
const isInRange = (min, max) => {
  /**
   *
   * @param {number} value
   * @returns boolean
   */
  function check(value) {
    return value >= min && value <= max;
  }
  return refine(number(), 'range', check);
};

const BuildTypeStruct = object({
  id: isInRange(10, 64),
  features: optional(unique(array(string()))),
  env: optional(EnvArrayStruct),
  isPrerelease: optional(boolean()),
  manifestOverrides: union([string(), literal(false)]),
  buildNameOverride: union([string(), literal(false)]),
});

const CopyAssetStruct = object({ src: string(), dest: string() });
const ExclusiveIncludeAssetStruct = coerce(
  object({ exclusiveInclude: string() }),
  string(),
  (exclusiveInclude) => ({ exclusiveInclude }),
);
const AssetStruct = union([CopyAssetStruct, ExclusiveIncludeAssetStruct]);

const FeatureStruct = object({
  env: optional(EnvArrayStruct),
  // TODO(ritave): Check if the paths exist
  assets: optional(array(AssetStruct)),
});

const FeaturesStruct = refine(
  record(
    string(),
    coerce(FeatureStruct, nullable(never()), () => ({})),
  ),
  'feature definitions',
  function* (value) {
    let isValid = true;

    const definitions = new Set();

    for (const feature of Object.values(value)) {
      for (const env of feature?.env ?? []) {
        if (typeof env !== 'string') {
          if (definitions.has(env.key)) {
            isValid = false;
            yield `Multiple defined features have a definition of "${env}" env variable, resulting in a conflict`;
          }
          definitions.add(env.key);
        }
      }
    }
    return isValid;
  },
);

const BuildTypesStruct = refine(
  object({
    default: string(),
    buildTypes: record(string(), BuildTypeStruct),
    features: FeaturesStruct,
    env: EnvArrayStruct,
  }),
  'BuildTypes',
  (value) => {
    if (!Object.keys(value.buildTypes).includes(value.default)) {
      return `Default build type "${value.default}" does not exist in builds declarations`;
    }
    return true;
  },
);

/**
 * Loads definitions of build type and what they are composed of.
 *
 * @returns {import('superstruct').Infer<typeof BuildTypesStruct>}
 */
function loadBuildTypesConfig() {
  if (cachedBuildTypes !== null) {
    return cachedBuildTypes;
  }
  const buildsData = yaml.parse(fs.readFileSync(BUILDS_YML_PATH, 'utf8'));
  const [err, result] = validate(buildsData, BuildTypesStruct, {
    coerce: true,
  });
  if (err !== undefined) {
    throw new AssertionError({
      message: constructFailureMessage(err),
    });
  }
  cachedBuildTypes = result;
  return cachedBuildTypes;
}

/**
 * Creates a user readable error message about parse failure.
 *
 * @param {import('superstruct').StructError} structError
 * @returns {string}
 */
function constructFailureMessage(structError) {
  return `Failed to parse builds.yml
  -> ${structError
    .failures()
    .map(
      (failure) =>
        `${failure.message} (${BUILDS_YML_PATH}:.${failure.path.join('/')})`,
    )
    .join('\n  -> ')}
`;
}

module.exports = { loadBuildTypesConfig };
