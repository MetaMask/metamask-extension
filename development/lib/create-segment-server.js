const http = require('http');

/**
 * This is the default error handler to be used by this mock segment server.
 * It will print the error to the console and exit the process.
 *
 * @param {Error} error - The server error
 */
function defaultOnError(error) {
  console.log(error);
  process.exit(1);
}

/**
 * @typedef {import('http').IncomingMessage} IncomingMessage
 * @typedef {import('http').ServerResponse} ServerResponse
 */

/**
 * This function handles requests for the mock Segment server
 *
 * @typedef {(request: IncomingMessage, response: ServerResponse, metricEvents: Array<object>) => void} MockSegmentRequestHandler
 */

/**
 * Creates a HTTP server that acts as a fake version of the Segment API.
 * The bevahiour is rudimentary at the moment - it returns HTTP 200 in response
 * to every request. The only function this serves is to spy on requests sent to
 * this server, and to parse the request payloads as Segment batch events.
 *
 * @param {MockSegmentRequestHandler} onRequest - A callback for each request the server receives.
 * @param {(error: Error) => void} [onError] - A callback for server error events
 */
function createSegmentServer(onRequest, onError = defaultOnError) {
  const server = http.createServer(async (request, response) => {
    const chunks = [];

    request.on('data', (chunk) => {
      chunks.push(chunk);
    });

    await new Promise((resolve) => {
      request.on('end', () => {
        resolve();
      });
    });

    // respond to preflight request
    if (request.method === 'OPTIONS') {
      response.setHeader('Access-Control-Allow-Origin', '*');
      response.setHeader('Access-Control-Allow-Methods', '*');
      response.setHeader('Access-Control-Allow-Headers', '*');
      response.statusCode = 200;
      response.end();
      return;
    }

    let metricEvents = [];
    if (chunks.length) {
      const body = Buffer.concat(chunks).toString();
      const segmentPayload = JSON.parse(body);
      metricEvents = segmentPayload.batch;
    }

    onRequest(request, response, metricEvents);
  });

  server.on('error', onError);

  return {
    start: async (port) => {
      await new Promise((resolve, reject) => {
        server.listen(port, (error) => {
          if (error) {
            return reject(error);
          }
          return resolve();
        });
      });
    },
    stop: async () => {
      await new Promise((resolve, reject) => {
        server.close((error) => {
          if (error) {
            return reject(error);
          }
          return resolve();
        });
        // We need to close all connections to stop the server quickly
        // Otherwise it takes a few seconds for it to close
        server.closeAllConnections();
      });
    },
  };
}

module.exports = { createSegmentServer };
