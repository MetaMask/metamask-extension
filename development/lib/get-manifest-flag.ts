import path from 'node:path';
import fs from 'node:fs/promises';
import { promisify } from 'node:util';
import { exec as callbackExec } from 'node:child_process';

import { hasProperty } from '@metamask/utils';
import { merge } from 'lodash';

import type { ManifestFlags } from '../../app/scripts/lib/manifestFlags';

const exec = promisify(callbackExec);
const PR_BODY_FILEPATH = path.resolve(
  __dirname,
  '..',
  '..',
  'changed-files',
  'pr-body.txt',
);

/**
 * Search a string for `flags = {...}` and return ManifestFlags if it exists
 *
 * @param str - The string to search
 * @param errorType - The type of error to log if parsing fails
 * @returns The ManifestFlags object if valid, otherwise undefined
 */
function regexSearchForFlags(str: string, errorType: string): ManifestFlags {
  // Search str for `flags = {...}`
  const flagsMatch = str.match(/flags\s*=\s*(\{.*\})/u);

  if (flagsMatch) {
    try {
      // Get 1st capturing group from regex
      return JSON.parse(flagsMatch[1]);
    } catch (error) {
      console.error(
        `Error parsing flags from ${errorType}, ignoring flags\n`,
        error,
      );
    }
  }

  return {};
}

/**
 * Get flags from the GitHub PR body if they are set
 *
 * To use this feature, add a line to your PR body like:
 * `flags = {"sentry": {"tracesSampleRate": 0.1}}`
 * (must be valid JSON)
 *
 * @returns Any manifest flags found in the PR body
 */
async function getFlagsFromPrBody(): Promise<ManifestFlags> {
  let body: string;
  try {
    body = await fs.readFile(PR_BODY_FILEPATH, 'utf8');
  } catch (error) {
    if (
      error instanceof Error &&
      hasProperty(error, 'code') &&
      error.code === 'ENOENT'
    ) {
      console.debug('No pr-body.txt, ignoring flags');
      return {};
    }
    throw error;
  }

  return regexSearchForFlags(body, 'PR body');
}

/**
 * Get flags from the Git message if they are set
 *
 * To use this feature, add a line to your commit message like:
 * `flags = {"sentry": {"tracesSampleRate": 0.1}}`
 * (must be valid JSON)
 *
 * @returns Any manifest flags found in the commit message
 */
async function getFlagsFromGitMessage(): Promise<ManifestFlags> {
  const gitMessage = (await exec(`git show --format='%B' --no-patch "HEAD"`))
    .stdout;

  return regexSearchForFlags(gitMessage, 'git message');
}

/**
 * Get any manifest flags found in the PR body and git message.
 *
 * @returns Any manifest flags found
 */
export async function fetchManifestFlagsFromPRAndGit(): Promise<ManifestFlags> {
  const [prBodyFlags, gitMessageFlags] = await Promise.all([
    getFlagsFromPrBody(),
    getFlagsFromGitMessage(),
  ]);

  return merge(prBodyFlags, gitMessageFlags);
}
