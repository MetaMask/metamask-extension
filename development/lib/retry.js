/**
 * Re-runs the given function until it returns a resolved promise or the number
 * of retries is exceeded, whichever comes first (with an optional delay in
 * between retries).
 *
 * @param {object} args - A set of arguments and options.
 * @param {number} args.retries - The maximum number of times to re-run the
 * function on failure.
 * @param {number} [args.delay] - The amount of time (in milliseconds) to wait in
 * between retries. (Default: 0)
 * @param {string} [args.rejectionMessage] - The message for the rejected promise
 * this function will return in the event of failure. (Default: "Retry limit
 * reached")
 * @param {boolean} [args.stopAfterOneFailure] - Retries until the function fails.
 * @param {Function} functionToRetry - The function that is run and tested for
 * failure.
 * @returns {Promise<* | null | Error>} a promise that either resolves with one of the following:
 * - If successful, resolves with the return value of functionToRetry.
 * - If functionToRetry fails while stopAfterOneFailure is true, resolves with null.
 * - Otherwise it is rejected with rejectionMessage.
 */
async function retry(
  {
    retries,
    delay = 0,
    rejectionMessage = 'Retry limit reached',
    stopAfterOneFailure = false,
  },
  functionToRetry,
) {
  let attempts = 0;
  while (attempts <= retries) {
    if (attempts > 0 && delay > 0) {
      await new Promise((resolve) => setTimeout(resolve, delay));
    }

    try {
      const result = await functionToRetry();
      if (!stopAfterOneFailure) {
        return result;
      }
    } catch (error) {
      if (error.message === "Exited with code '1'") {
        console.log("retry() received: Exited with code '1'");
      } else {
        console.error('error caught in retry():', error);
      }

      if (stopAfterOneFailure) {
        throw new Error('Test failed. No more retries will be performed');
      }

      if (attempts < retries) {
        console.log('Ready to retry() again');
      }
    } finally {
      attempts += 1;
    }
  }

  if (stopAfterOneFailure) {
    return null;
  }

  throw new Error(rejectionMessage);
}

module.exports = { retry };
