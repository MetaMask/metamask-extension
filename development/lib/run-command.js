const fs = require('fs');
const { spawn } = require('node:child_process');

/**
 * Run a command to completion using the system shell.
 *
 * This will run a command with the specified arguments, and resolve when the
 * process has exited. The STDOUT stream is monitored for output, which is
 * returned after being split into lines. All output is expected to be UTF-8
 * encoded, and empty lines are removed from the output.
 *
 * Anything received on STDERR is assumed to indicate a problem, and is tracked
 * as an error.
 *
 * @param {string} command - The command to run
 * @param {Array<string>} [args] - The arguments to pass to the command
 * @returns {Promise<Array<string>>} Lines of output received via STDOUT
 */
async function runCommand(command, args) {
  const output = [];
  let mostRecentError;
  let errorSignal;
  let errorCode;
  const internalError = new Error('Internal');
  try {
    await new Promise((resolve, reject) => {
      const childProcess = spawn(command, args, { encoding: 'utf8' });
      childProcess.stdout.setEncoding('utf8');
      childProcess.stderr.setEncoding('utf8');

      childProcess.on('error', (error) => {
        mostRecentError = error;
      });

      childProcess.stdout.on('data', (message) => {
        const nonEmptyLines = message.split('\n').filter((line) => line !== '');
        output.push(...nonEmptyLines);
      });

      childProcess.stderr.on('data', (message) => {
        mostRecentError = new Error(message.trim());
      });

      childProcess.once('exit', (code, signal) => {
        if (code === 0) {
          return resolve();
        }
        errorCode = code;
        errorSignal = signal;
        return reject(internalError);
      });
    });
  } catch (error) {
    /**
     * The error is re-thrown here in an `async` context to preserve the stack trace. If this was
     * was thrown inside the Promise constructor, the stack trace would show a few frames of
     * Node.js internals then end, without indicating where `runCommand` was called.
     */
    if (error === internalError) {
      let errorMessage;
      if (errorCode !== null && errorSignal !== null) {
        errorMessage = `Terminated by signal '${errorSignal}'; exited with code '${errorCode}'`;
      } else if (errorSignal !== null) {
        errorMessage = `Terminaled by signal '${errorSignal}'`;
      } else if (errorCode === null) {
        errorMessage = 'Exited with no code or signal';
      } else {
        errorMessage = `Exited with code '${errorCode}'`;
      }
      const improvedError = new Error(errorMessage);
      if (mostRecentError) {
        improvedError.cause = mostRecentError;
      }
      throw improvedError;
    }
  }
  return output;
}

/**
 * Run a command to using the system shell.
 *
 * This will run a command with the specified arguments, and resolve when the
 * process has exited. The STDIN, STDOUT and STDERR streams are inherited,
 * letting the command take over completely until it completes. The success or
 * failure of the process is determined entirely by the exit code; STDERR
 * output is not used to indicate failure.
 *
 * @param {string} command - The command to run
 * @param {Array<string>} [args] - The arguments to pass to the command
 * @param {string} output - The output file to write to
 */
async function runInShell(command, args, output) {
  let errorSignal;
  let errorCode;
  const internalError = new Error('Internal');
  try {
    await new Promise((resolve, reject) => {
      const childProcess = spawn(command, args, { shell: true });
      childProcess.stdout.setEncoding('utf8');
      childProcess.stderr.setEncoding('utf8');
      childProcess.stdout.pipe(process.stdout);
      childProcess.stderr.pipe(process.stderr);
      if (output) {
        childProcess.stdout.pipe(fs.createWriteStream(output));
      }

      childProcess.once('exit', (code, signal) => {
        if (code === 0) {
          return resolve();
        }
        errorCode = code;
        errorSignal = signal;
        return reject(internalError);
      });
    });
  } catch (error) {
    /**
     * The error is re-thrown here in an `async` context to preserve the stack trace. If this was
     * was thrown inside the Promise constructor, the stack trace would show a few frames of
     * Node.js internals then end, without indicating where `runInShell` was called.
     */
    if (error === internalError) {
      let errorMessage;
      if (errorCode !== null && errorSignal !== null) {
        errorMessage = `Terminated by signal '${errorSignal}'; exited with code '${errorCode}'`;
      } else if (errorSignal !== null) {
        errorMessage = `Terminaled by signal '${errorSignal}'`;
      } else if (errorCode === null) {
        errorMessage = 'Exited with no code or signal';
      } else {
        errorMessage = `Exited with code '${errorCode}'`;
      }
      const improvedError = new Error(errorMessage);
      throw improvedError;
    }
  }
}

module.exports = { runCommand, runInShell };
