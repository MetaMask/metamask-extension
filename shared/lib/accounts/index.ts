export * from './accounts';
export * from './bitcoin-wallet-snap';
export * from './snaps';
export * from './solana-wallet-snap';
