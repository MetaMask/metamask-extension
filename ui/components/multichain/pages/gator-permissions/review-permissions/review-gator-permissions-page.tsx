import React from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { Hex } from '@metamask/utils';
import {
  ButtonIcon,
  ButtonIconSize,
  IconName,
  Text,
  IconColor,
  TextAlign,
  TextVariant,
} from '@metamask/design-system-react';
import { Header, Page } from '../../page';
import { BackgroundColor } from '../../../../../helpers/constants/design-system';
import { useI18nContext } from '../../../../../hooks/useI18nContext';
import { extractNetworkName } from '../helper';
import { getMultichainNetworkConfigurationsByChainId } from '../../../../../selectors';

export const ReviewGatorPermissionsPage = () => {
  const t = useI18nContext();
  const history = useHistory();
  const { chainId } = useParams();
  const [, evmNetworks] = useSelector(
    getMultichainNetworkConfigurationsByChainId,
  );
  const getNetworkNameForChainId = () => {
    if (!chainId) {
      return t('unknownNetworkForGatorPermissions');
    }
    const networkNameKey = extractNetworkName(evmNetworks, chainId as Hex);
    const networkName = t(networkNameKey);

    // If the translation key doesn't exist (returns the same key), fall back to the full network name
    if (!networkName || networkName === networkNameKey) {
      return extractNetworkName(evmNetworks, chainId as Hex, true);
    }

    return networkName;
  };

  return (
    <Page
      className="main-container"
      data-testid="review-gator-permissions-page"
      key="review-gator-permissions-page"
    >
      <Header
        backgroundColor={BackgroundColor.backgroundDefault}
        startAccessory={
          <ButtonIcon
            ariaLabel={t('back')}
            iconName={IconName.ArrowLeft}
            className="connections-header__start-accessory"
            color={IconColor.IconDefault}
            onClick={() => history.goBack()}
            size={ButtonIconSize.Sm}
          />
        }
      >
        <Text
          variant={TextVariant.HeadingMd}
          textAlign={TextAlign.Center}
          data-testid="review-gator-permissions-page-title"
        >
          {getNetworkNameForChainId()}
        </Text>
      </Header>
    </Page>
  );
};
