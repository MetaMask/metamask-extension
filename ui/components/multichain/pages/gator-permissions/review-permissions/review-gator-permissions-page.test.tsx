import React from 'react';
import { renderWithProvider } from '../../../../../../test/lib/render-helpers';
import mockState from '../../../../../../test/data/mock-state.json';
import configureStore from '../../../../../store/store';
import { ReviewGatorPermissionsPage } from './review-gator-permissions-page';

const store = configureStore({
  ...mockState,
  metamask: {
    ...mockState.metamask,
  },
});

describe('Review Gator Permissions Page', () => {
  describe('render', () => {
    it('renders correctly', () => {
      const { container, getByTestId } = renderWithProvider(
        <ReviewGatorPermissionsPage />,
        store,
      );
      expect(container).toMatchSnapshot();

      expect(getByTestId('review-gator-permissions-page')).toBeInTheDocument();
    });
  });
});
