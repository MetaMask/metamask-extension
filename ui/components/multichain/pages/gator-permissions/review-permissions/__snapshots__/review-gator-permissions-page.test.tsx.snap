// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review Gator Permissions Page render renders correctly 1`] = `
<div>
  <div
    class="mm-box multichain-page multichain-page--has-app-header mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--width-full mm-box--height-full mm-box--background-color-background-default"
    data-testid="multichain-page"
  >
    <div
      class="mm-box multichain-page__inner-container main-container mm-box--display-flex mm-box--flex-direction-column mm-box--width-full mm-box--height-full mm-box--background-color-background-default"
      data-testid="review-gator-permissions-page"
    >
      <div
        class="mm-box mm-header-base multichain-page-header mm-box--padding-4 mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--width-full mm-box--background-color-background-default"
      >
        <div
          class="mm-box"
          style="min-width: 0px;"
        >
          <button
            aria-label="Back"
            class="inline-flex items-center justify-center p-0 h-6 w-6 rounded bg-transparent hover:bg-hover active:bg-pressed text-icon-default connections-header__start-accessory"
            color="text-icon-default"
          >
            <svg
              class="inline-block w-4 h-4 text-inherit"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m15.112 21.886-10-10 10-10 1.775 1.775-8.225 8.224 8.225 8.226z"
              />
            </svg>
          </button>
        </div>
        <div
          class="mm-box"
        >
          <p
            class="mm-box mm-text mm-text--body-md-bold mm-text--ellipsis mm-text--text-align-center mm-box--padding-inline-start-2 mm-box--padding-inline-end-2 mm-box--display-block mm-box--color-text-default"
          >
            <h3
              class="text-default text-s-heading-md leading-s-heading-md tracking-s-heading-md md:text-l-heading-md md:leading-l-heading-md md:tracking-l-heading-md font-bold font-default text-center"
              data-testid="review-gator-permissions-page-title"
            >
              Unknown network
            </h3>
          </p>
        </div>
      </div>
      <div
        class="mm-box multichain-page-content mm-box--padding-0 mm-box--padding-top-0 mm-box--display-flex mm-box--flex-direction-column mm-box--width-full mm-box--height-full"
        style="overflow: auto;"
      >
        <div
          class="flex flex-col gap-2 justify-center p-4"
          data-testid="no-connections"
        >
          <p
            class="text-default text-s-body-md leading-s-body-md tracking-s-body-md md:text-l-body-md md:leading-l-body-md md:tracking-l-body-md font-regular font-default text-center"
          >
            Nothing to see here
          </p>
          <p
            class="text-alternative text-s-body-md leading-s-body-md tracking-s-body-md md:text-l-body-md md:leading-l-body-md md:tracking-l-body-md font-regular font-default text-center"
          >
            This is where you can see the permissions you've given to installed Snaps or connected sites.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
